rr_iarrc/src/cone_detection/cone_height_detector.cpp

```diff
@@ -0,0 +1,379 @@
+#include <boost/algorithm/string.hpp>
+#include <boost/math/constants/constants.hpp>
+#include <algorithm>
+#include <pcl_ros/point_cloud.h>
+#include <iostream>
+#include <stdlib.h>
+#include <stdio.h>
+#include <sensor_msgs/CameraInfo.h>
+#include <std_msgs/Float32.h>
+
+using namespace std;
+
+// Image Processing
+void blockEnvironment(const cv::Mat&);
+cv::Mat find_orange_in_ROI(const cv::Mat&);
+cv::Mat cutSmall(const cv::Mat&, int);
+cv::Mat canny_cut_bodies(const cv::Mat&, const cv::Mat&);
+cv::Mat find_unique_centers(const cv::Mat&, double);
+cv::Mat perform_watershed(const cv::Mat&, const cv::Mat&, const cv::Mat&);
+cv::Mat draw_box_and_calc_dist(const cv::Mat&, const cv::Mat&);
+cv::Mat kernel(int, int);
+void publishImage(ros::Publisher, cv::Mat, std::string);
+
+ros::Publisher pub_debug_pos, pub_debug_mask, pub_pointcloud, pub_closest_point;
+int blockSky_height, blockWheels_height, blockBumper_height;
+typedef cv::Mat Mat;
+typedef cv::Point Point;
+int low_H, high_H, low_S, low_V;
+int canny_cut_min_threshold, minimum_area_cut;
+double percent_max_distance_transform;
+
+
+// PointCloud Processing
+pcl::PointCloud<pcl::PointXYZ> draw_cone_circles();
+void drawCircle(pcl::PointCloud<pcl::PointXYZ>&, pcl::PointXYZ, double, int);
+void fovCallback(const sensor_msgs::CameraInfoConstPtr& msg);
+void loadCameraFOV(ros::NodeHandle& nh);
+void publish_closest_point();
+
+cv_bridge::CvImagePtr cv_ptr;
+typedef pcl::PointCloud<pcl::PointXYZ> PointCloud;
+double fy, real_cone_height, real_cone_radius;
+double camera_fov_horizontal;
+cv::Size imageSize;
+bool fov_callback_called;
+double angle_constant;
+vector<pcl::PointXYZ> cone_points;
+
+
+void publish_closest_point() {
+    if (cone_points.size() != 0){
+        double min = cone_points[0].x;
+        for (pcl::PointXYZ cone_point : cone_points) {
+            if (cone_point.x < min) {
+                min = cone_point.x;
+            }
+        }
+        std_msgs::Float32 msg;
+        msg.data = static_cast<float>(min);
+        pub_closest_point.publish(msg);
+    }
+}
+
+
+void fovCallback(const sensor_msgs::CameraInfoConstPtr& msg) {
+    double fx = msg->P[0]; //horizontal focal length of rectified image, in px
+//    fy = msg->P[5]; //vertical focal length
+    imageSize = cv::Size(msg->width, msg->height);
+    camera_fov_horizontal = 2 * atan2(imageSize.width, 2*fx);
+    fov_callback_called = true;
+    angle_constant = camera_fov_horizontal / 2.0 /imageSize.width;
+}
+
+
+void loadCameraFOV(ros::NodeHandle& nh) {
+    auto infoSub = nh.subscribe("/camera/camera_info", 1, fovCallback);
+
+    fov_callback_called = false;
+    while (!fov_callback_called) {
+        ros::spinOnce();
+        ros::Duration(1).sleep();
+    }
+    ROS_INFO("Using horizontal FOV %f ", camera_fov_horizontal);
+}
+
+
+void publishPointCloud(pcl::PointCloud<pcl::PointXYZ> &cloud) {
+    sensor_msgs::PointCloud2 outmsg;
+    pcl::toROSMsg(cloud, outmsg);
+    outmsg.header.frame_id = "base_footprint";
```

Comment:
Is the point cloud output actually relative to the base_footprint? I think it is relative to the camera link.